### DISCUSSION TITLE
Failed to change DNS via resolvconf

### DISCUSSION THREAD

**Discussion by ToolmanP** [strongswan/strongswan#2085] [Q&A]
So i'm using VPN provided from my university and it requires a specific DNS modification.

Since resolve plugin from what i'm concerned, 
>strongSwan provides support for installing name servers via the [resolvconf](http://en.wikipedia.org/wiki/Resolvconf) framework (or e.g. the implementation provided by systemd’s resolvectl). If resolvconf is found or its path is explicitly configured, the plugin automatically invokes it appropriately instead of modifying resolv.conf directly.

First, I'm using resolvconf provided by the systemd-resolved which is just a symlink to resolvctl with little luck. I think the configuration details have changed since some newer version of systemd. The issue happens here with this issue 

> resolvconf: Failed to set DNS configuration: Link lo is loopback device

Then, i gave a try with open-resolv also with little luck, with something like

> resolvconf: /lib/resolvconf/libc: line 246: /etc/resolv.conf: Permission denied

So i guess the process is using ipsec role to launch resolvconf? If so, how can i manage to get around with that?

**Comment by qpkorr** [comment]
Several of us are having a similar issue with "lo is loopback device" - over here
https://bugs.launchpad.net/ubuntu/+source/systemd/+bug/2067897
(once we got past another issue, networkd service not running after reboot).

I'm just wondering, is lo.ipsec the best default interface to use, given this issue with loopback device:
#: john@j4lt:~ ; ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00

Should we be adding some other interface, perhaps via a hook script eg running
sudo ip link add vpn_interface_name type dummy

and then set vpn_interface_name (ie whatever name is preferred) as the default interface? (adding the .ipsec protocol part where appropriate?)  

Basically, is there a better way to solve this that doesn't require users editing /etc/strongswan.d/charon/resolve.conf ?
Thanks in advance for your help!

**Reply by tobiasbrunner** [reply]
> I'm just wondering, is lo.ipsec the best default interface to use, given this issue with loopback device:

What interface would you suggest that is available on basically all systems? And if you refer to `systemd-resolved` rejecting it, then ask its devs why they do that (or why they tie DNS servers to interfaces in the first place).

But you already have the option to configure an interface that's available on your system and is accepted by `systemd-resolved`. What's the problem with that? And I guess you could also install `resolvconf` or `openresolv` to get a version of `resolvconf` that doesn't require config changes.

And sure, you could use XFRM interfaces if you wanted to (it's what the current version of the strongSwan NM plugin does), but that requires several other config settings, in particular, if the remote traffic selectors overlap with the server's address. (Note that using dummy interfaces could also be problematic, as we saw with NM when it uses `dnsmasq`, which actually binds to the interface to send the DNS requests.)

**Reply by cureton** [reply]
HI Tobias, I was helping john debug this yesterday.   however going forward it may be useful for ipsec to establish its own private "lo" interface for the purpose of storing and removing DNS configurations. 

https://baturin.org/docs/iproute2/#ip-link

Reading this reference on iproute2 a dummy interface is merely a private  'lo" interface

Bringing up your own private lo interface allow the ipsec software to associate DNS  servers that are advertised and accessible via the VPN link and even domain search paths to the private lo interfaace.  This also allows ipsec to destroy the interface thus  negating DNS servers and domain search paths associates with that link on the link pull down, without have to be mindful of configurations that other software may have applied and carefully un-pick the setting applied by charron. 

In our case the work around is to add creation of the private lo interface in a pre-ipsec up script,  specify the same private interface in the charon config and then destroy the private lo interface in ipsec down script. 

These private interface should ideally be on a per-vpn connection identifier so each connection can manage its connection related host configuration simply. 

Trying to manage resovl.conf n a global sense has always been problematic and this makes the management of that resource at a link level. 

I don't think this is a case of an bug per-se in either Strongswan or more specifically swanctl  or systemd-resolvctl, but just a matter of interoperability between the two. 

There is a lot of legacy deployments  on top of IPSEC that would need to be considered regarding a change like this and no-one is more equipped that you to understand these considerations, however we though it good to share our solution in case it swayed further development of Strongwan in the future and made the management of resolv.conf information simpler going forward

**Reply by qpkorr** [reply]
Thanks very much Tobias (and in case it wasn't clear, I'm the "john" that cureton mentioned).

You asked lots of good questions - which you no doubt know the answers to far better than me!
"What interface would you suggest that is available on basically all systems?" My thought would be to add one as cureton suggested - but only if dnsmasq can be updated to work with this?

"if you refer to systemd-resolved rejecting it, then ask its devs why they do that" - they've not yet addressed the ticket in which two of us are asking about this issue (linked in my first message)

"or why they tie DNS servers to interfaces in the first place" - I could guess "security" - but you'll know far better than me I imagine.

"But you already have the option to configure an interface that's available on your system and is accepted by systemd-resolved. What's the problem with that?" - not much - it felt like a regression from previous functionality (using non-systemd resolvconf) - and the interfaces I'm using change a bit depending if I'm on wifi or ethernet (over usb-c).

"I guess you could also install resolvconf or openresolv" - I did that with ubuntu 22, but it seems neither is available from Ubuntu for 23 or 24, so again, it didn't feel like the optimal solution.

Thanks for the suggestion re XFRM interfaces - I'm yet to explore that option.

Thanks for the comments re dummy interfaces - I read some discussions/issues about them.

It certainly sounds like a very complicated and frustrating domain - once again, thanks so much for all your time/efforts with it.  Unsurprisingly I've still got no idea how things should ideally work - though it still feels like the might be scope for improvements, somewhere, since systemd's changes.

PS - I also hit a problem, using a dummy interface for the vpn, where a server had a different address on the public internet vs on my vpn.  With the old resolvconf, bringing up the vpn automatically made me get the VPN-served ip address for that server - but using a dummy interface, I still got the public version.  If I set iface in /etc/strongswan.d/charon/resolve.conf to my ethernet interface, it replaces the public DNS and this problem goes away - but dropping the VPN deletes all DNS settings on that interface... the old resolvconf really was a lot more functional - as I've mentioned in the systemd bug!

**Comment by cureton** [comment]
I believe the last bit about split horizon dns can be resolved by using ~ceos.com.au as well as other corporate domains with a tilde as as  DOMAIN statement on the dummy interface that we bring up on IPsec tunnel establishment. thus this DNS server will preferred to resolve the specified domain.   also when this  then the dummy interface is dropped by IPsec going down then these network preferences are extinguished automatically too.  On 7 Jun 2024, at 11:30 AM, John Orr ***@***.***> wrote:﻿
Thanks very much Tobias (and in case it wasn't clear, I'm the "john" that cureton mentioned).
You asked lots of good questions - which you no doubt know the answers to far better than me!
"What interface would you suggest that is available on basically all systems?" My thought would be to add one as cureton suggested - but only if dnsmasq can be updated to work with this?
"if you refer to systemd-resolved rejecting it, then ask its devs why they do that" - they've not yet addressed the ticket in which two of us are asking about this issue (linked in my first message)
"or why they tie DNS servers to interfaces in the first place" - I could guess "security" - but you'll know far better than me I imagine.
"But you already have the option to configure an interface that's available on your system and is accepted by systemd-resolved. What's the problem with that?" - not much - it felt like a regression from previous functionality (using non-systemd resolvconf) - and the interfaces I'm using change a bit depending if I'm on wifi or ethernet (over usb-c).
"I guess you could also install resolvconf or openresolv" - I did that with ubuntu 22, but it seems neither is available from Ubuntu for 23 or 24, so again, it didn't feel like the optimal solution.
Thanks for the suggestion re XFRM interfaces - I'm yet to explore that option.
Thanks for the comments re dummy interfaces - I read some discussions/issues about them.
It certainly sounds like a very complicated and frustrating domain - once again, thanks so much for all your time/efforts with it.  Unsurprisingly I've still got no idea how things should ideally work - though it still feels like the might be scope for improvements, somewhere, since systemd's changes.
PS - I also hit a problem, using a dummy interface for the vpn, where a server had a different address on the public internet vs on my vpn.  With the old resolvconf, bringing up the vpn automatically made me get the VPN-served ip address for that server - but using a dummy interface, I still got the public version.  If I set iface in /etc/strongswan.d/charon/resolve.conf to my ethernet interface, it replaces the public DNS and this problem goes away - but dropping the VPN deletes all DNS settings on that interface... the old resolvconf really was a lot more functional - as I've mentioned in the systemd bug!

**Comment by ToolmanP** [comment]
Hey guys, some recent updates. I've made through it by deliberately enforcing manual DNS on my network connection and the problem magically went away. I just don't know what happens though.

**Comment by fnature** [comment]
Hi there, I have the same issue when starting my ipsec vpn. I'm lost with how to fix it permanently. Only workaround I found was to add DNS nameserver manually in my /etc/systemd/resolved.conf. But I keep having to restart my systemd-resolved. I'm not sure how to implement the dummy interface solution, and it appears to not be ideal as well ?
